File: common/src/main/java/de/hype/bingonet/environment/addonpacketconfig/AddonPacketUtils.java
package de.hype.bingonet.environment.addonpacketconfig;

import com.google.gson.Gson;
import de.hype.bingonet.client.common.chat.Chat;
import de.hype.bingonet.client.common.client.BingoNet;
import de.hype.bingonet.client.common.client.CustomGson;
import de.hype.bingonet.client.common.client.socketAddons.AddonHandler;

import java.util.function.Consumer;


public class AddonPacketUtils {
    public static final Gson gson = CustomGson.createNotPrettyPrinting();

    public static String parsePacketToJson(AbstractAddonPacket packet) {
        return gson.toJson(packet).replace("\n", "/n");
    }

    public static <T extends AbstractAddonPacket> void tryToProcessPacket(AddonPacket<T> addonPacket, String rawJson) {
        Class<T> clazz = addonPacket.getClazz();
        Consumer<T> consumer = addonPacket.getConsumer();
        T abstractPacket = gson.fromJson(rawJson.replace("/n", "\n"), clazz);
        consumer.accept(abstractPacket);
    }

    private static void showError(Throwable t, String errorMessage) {
        System.out.println(errorMessage + " because of: " + t.getClass().getSimpleName() + ":  " + t.getMessage());
        new Error(errorMessage, t).printStackTrace();
    }

    public static <T extends AbstractAddonPacket> T getAsPacket(String message, Class<T> clazz) {
        if (!message.contains(".")) return null;
        String packetName = message.split("\\.")[0];
        String rawJson = message.substring(packetName.length() + 1);
        if (!packetName.equals(clazz.getSimpleName())) {
            try {
                T parsedPacket = gson.fromJson(rawJson.replace("/n", "\n"), clazz);
                return parsedPacket;
            } catch (Throwable t) {
                showError(t, "Could not process Addon packet '" + packetName + "' from " + AddonEnvironmentPacketConfig.notEnviroment);
            }
        }
        String errorMessage = "Could not process Addon packet '" + packetName + "' from " + AddonEnvironmentPacketConfig.notEnviroment;

        showError(new APIException("Found unknown Addon packet: " + packetName + "'"), errorMessage);
        return null;
    }

    public static boolean isPacket(String message, Class<? extends AbstractAddonPacket> clazz) {
        if (!message.contains(".")) return false;
        String packetName = message.split("\\.")[0];
        if (packetName.equals(clazz.getSimpleName())) {
            return true;
        }
        return false;
    }

    public static boolean isPacket(String message) {
        if (!message.contains(".")) return false;
        String packetName = message.split("\\.")[0];
        for (Class<? extends AbstractAddonPacket> packetClass : AddonPacketManager.getAllPacketClasses()) {
            if (!packetName.equals(packetClass.getSimpleName())) {
                return true;
            }
        }
        return false;
    }

    public static <T extends AbstractAddonPacket> boolean handleIfPacket(AddonHandler connection, String message) {
        //Return = is Packet
        if (!message.contains(".")) return false;
        String packetName = message.split("\\.")[0];
        String rawJson = message.substring(packetName.length() + 1);
        AddonPacketManager manager = new AddonPacketManager(connection);
        for (AddonPacket<? extends AbstractAddonPacket> addonPacket : manager.getPackets()) {
            if (!packetName.equals(addonPacket.getClazz().getSimpleName())) continue;
            try {
                if (BingoNet.socketAddonConfig.addonDebug)
                    Chat.sendPrivateMessageToSelfDebug(packetName + ":" + rawJson);
                tryToProcessPacket(addonPacket, rawJson);
                return true;
            } catch (Throwable t) {
                showError(t, "Could not process Addon packet '" + packetName + "' from " + AddonEnvironmentPacketConfig.notEnviroment);
            }
        }
        String errorMessage = "Could not process Addon packet '" + packetName + "' from " + AddonEnvironmentPacketConfig.notEnviroment;

        showError(new APIException("Found unknown Addon packet: " + packetName + "'"), errorMessage);
        return false;
    }

    public static class APIException extends Error {

        public APIException(String errorMessage, Throwable t) {
            super(errorMessage, t);
        }

        public APIException(String errorMessage) {
            super(errorMessage);
        }
    }
}

File: common/src/main/kotlin/de/hype/bingonet/environment/addonpacketconfig/AddonPacketUtils.kt
package de.hype.bingonet.environment.addonpacketconfig

import com.google.gson.Gson
import de.hype.bingonet.client.common.chat.Chat
import de.hype.bingonet.client.common.client.BingoNet
import de.hype.bingonet.client.common.client.CustomGson
import de.hype.bingonet.client.common.client.socketAddons.AddonHandler

object AddonPacketUtils {
    val gson: Gson = CustomGson.createNotPrettyPrinting()

    fun parsePacketToJson(packet: AbstractAddonPacket?): String {
        return gson.toJson(packet).replace("\n", "/n")
    }

    fun <T : AbstractAddonPacket> tryToProcessPacket(addonPacket: AddonPacket<T>, rawJson: String) {
        val clazz = addonPacket.clazz
        val consumer = addonPacket.consumer
        val abstractPacket = gson.fromJson(rawJson.replace("/n", "\n"), clazz)
        consumer.accept(abstractPacket)
    }

    private fun showError(t: Throwable, errorMessage: String?) {
        println(errorMessage + " because of: " + t.javaClass.getSimpleName() + ":  " + t.message)
        Error(errorMessage, t).printStackTrace()
    }

    fun <T : AbstractAddonPacket?> getAsPacket(message: String, clazz: Class<T?>): T? {
        if (!message.contains(".")) return null
        val packetName = message.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0]
        val rawJson = message.substring(packetName.length + 1)
        if (packetName != clazz.getSimpleName()) {
            try {
                val parsedPacket = gson.fromJson<T?>(rawJson.replace("/n", "\n"), clazz)
                return parsedPacket
            } catch (t: Throwable) {
                showError(
                    t,
                    "Could not process Addon packet '" + packetName + "' from " + AddonEnvironmentPacketConfig.notEnviroment
                )
            }
        }
        val errorMessage =
            "Could not process Addon packet '" + packetName + "' from " + AddonEnvironmentPacketConfig.notEnviroment

        showError(APIException("Found unknown Addon packet: " + packetName + "'"), errorMessage)
        return null
    }

    fun isPacket(message: String, clazz: Class<out AbstractAddonPacket?>): Boolean {
        if (!message.contains(".")) return false
        val packetName = message.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0]
        if (packetName == clazz.getSimpleName()) {
            return true
        }
        return false
    }

    fun isPacket(message: String): Boolean {
        if (!message.contains(".")) return false
        val packetName = message.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0]
        for (packetClass in AddonPacketManager.Companion.allPacketClasses) {
            if (packetName != packetClass.getSimpleName()) {
                return true
            }
        }
        return false
    }

    fun <T : AbstractAddonPacket?> handleIfPacket(connection: AddonHandler?, message: String): Boolean {
        //Return = is Packet
        if (!message.contains(".")) return false
        val packetName = message.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0]
        val rawJson = message.substring(packetName.length + 1)
        val manager = AddonPacketManager(connection)
        for (addonPacket in manager.packets) {
            if (packetName != addonPacket.clazz.simpleName) continue
            try {
                if (BingoNet.socketAddonConfig.addonDebug) Chat.sendPrivateMessageToSelfDebug(packetName + ":" + rawJson)
                AddonPacketUtils.tryToProcessPacket(addonPacket, rawJson)
                return true
            } catch (t: Throwable) {
                showError(
                    t,
                    "Could not process Addon packet '" + packetName + "' from " + AddonEnvironmentPacketConfig.notEnviroment
                )
            }
        }
        val errorMessage =
            "Could not process Addon packet '" + packetName + "' from " + AddonEnvironmentPacketConfig.notEnviroment

        showError(APIException("Found unknown Addon packet: $packetName'"), errorMessage)
        return false
    }

    class APIException : Error {
        constructor(errorMessage: String, t: Throwable) : super(errorMessage, t)

        constructor(errorMessage: String) : super(errorMessage)
    }
}
